cns/networkcontainers/networkcontainers_linux.go

```diff
@@ -3,20 +3,180 @@
+	"fmt"
+	"go/types"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/log"
+	"github.com/containernetworking/cni/libcni"
+	"github.com/containernetworking/cni/pkg/invoke"
+)
+
+const (
+	VersionStr string = "cniVersion"
+	PluginsStr string = "plugins"
+	NameStr    string = "name"
+)
 
 func createOrUpdateInterface(createNetworkContainerRequest cns.CreateNetworkContainerRequest) error {
-	return nil
+
+	if createNetworkContainerRequest.NetworkContainerType == cns.WebApps {
+		log.Printf("[Azure CNS] Operation not supported for WebApps Orchestrator.")
+		return nil
+	}
+
+	exists, _ := interfaceExists(createNetworkContainerRequest.NetworkContainerid)
+	if !exists {
+		log.Printf("[Azure CNS] Only Update Operation is supported.")
+		return nil
+	}
+
+	return createOrUpdateWithOperation(createNetworkContainerRequest, "UPDATE")
 }
 
 func setWeakHostOnInterface(ipAddress string) error {
 	return nil
 }
 
 func createOrUpdateWithOperation(createNetworkContainerRequest cns.CreateNetworkContainerRequest, operation string) error {
+	log.Printf("[Azure CNS] createOrUpdateWithOperation called with operation type %v", operation)
+
+	if _, err := os.Stat("/opt/cni/bin/azure-vnet"); err != nil {
+		if os.IsNotExist(err) {
+			return errors.New("[Azure CNS] Unable to find azure-vnet under /opt/cni/bin/. Cannot continue")
+		}
+	}
+
+	if createNetworkContainerRequest.IPConfiguration.IPSubnet.IPAddress == "" {
+		return errors.New("[Azure CNS] IPAddress in IPConfiguration of createNetworkContainerRequest is nil")
+	}
+
+	var podInfo cns.KubernetesPodInfo
+	err := json.Unmarshal(createNetworkContainerRequest.OrchestratorContext, &podInfo)
+	if err != nil {
+		log.Printf("[Azure CNS] Unmarshalling %s failed with error %v", createNetworkContainerRequest.NetworkContainerType, err)
+		return err
+	}
+
+	log.Printf("[Azure CNS] Pod info %v", podInfo)
+
+	// How to construct net namespace and container Id?
+	rt, err := buildCNIRuntimeConf(podInfo.PodName, podInfo.PodNamespace, "", "", createNetworkContainerRequest.NetworkContainerid)
+	if err != nil {
+		log.Printf("[Azure CNS] Failed to build runtime configuration with error %v", err)
+		return err
+	}
+
+	log.Printf("[Azure CNS] run time conf info %v", rt)
+
+	// Hardcoded path ?
+	netConf, err := getNetworkConf("/etc/cni/net.d/10-azure.conflist")
+	if err != nil {
+		log.Printf("[Azure CNS] Failed to build network configuration with error %v", err)
+		return err
+	}
+
+	log.Printf("[Azure CNS] network configuration info %v", string(netConf))
+
+	err = updateNetwork(rt, netConf)
+	if err != nil {
+		log.Printf("[Azure CNS] Failed to update network with error %v", err)
+		return err
+	}
+
 	return nil
 }
 
 func deleteInterface(networkContainerID string) error {
 	return nil
 }
+
+func buildCNIRuntimeConf(podName string, podNs string, podSandboxId string, podNetnsPath string, interfaceName string) (*libcni.RuntimeConf, error) {
```

Comment:
Instead of using the libcni.runtime we may need to modify it to work with our cni (azure-vnet); so that we may pass extra args like enableSnatOnHost, AdditionalArgs, multiTenancy, etc.

Reply:
As per our F2F conversation, this configuration represents command line argument and environment variables.